crates/client/rpc/src/versions/v0_8_0/methods/ws/lib.rs

```diff
@@ -0,0 +1,370 @@
+                    return Err(StarknetWsApiError::TooManyBlocksBack.into());
+                }
+
+                block_n
+            }
+            starknet_core::types::BlockId::Hash(block_hash) => {
+                let err = || format!("Failed to retrieve block info at hash {block_hash:#x}");
+                let block_latest = self
+                    .backend
+                    .get_block_n(&mp_block::BlockId::Tag(mp_block::BlockTag::Latest))
+                    .or_else_internal_server_error(err)?
+                    .ok_or(StarknetWsApiError::BlockNotFound)?;
+
+                let block_n = self
+                    .backend
+                    .get_block_n(&block_id)
+                    .or_else_internal_server_error(err)?
+                    .ok_or_else_internal_server_error(err)?;
+
+                if block_n < block_latest.saturating_sub(BLOCK_PAST_LIMIT) {
+                    return Err(StarknetWsApiError::TooManyBlocksBack.into());
+                }
+
+                block_n
+            }
+            starknet_core::types::BlockId::Tag(starknet_core::types::BlockTag::Latest) => self
+                .backend
+                .get_latest_block_n()
+                .or_internal_server_error("Failed to retrieve block info for latest block")?
+                .ok_or_internal_server_error("Failed to retrieve block info for latest block")?,
+            starknet_core::types::BlockId::Tag(starknet_core::types::BlockTag::Pending) => {
+                return Err(StarknetWsApiError::Pending.into());
+            }
+        };
+
+        let mut rx = self.backend.subscribe_block_info();
+        for n in block_n.. {
+            if sink.is_closed() {
+                return Ok(());
+            }
+
+            let block_info = match self.backend.get_block_info(&mp_block::BlockId::Number(n)) {
+                Ok(Some(block_info)) => {
+                    let err = || format!("Failed to retrieve block info for block {n}");
+                    block_info.as_nonpending_owned().ok_or_else_internal_server_error(err)?
+                }
+                Ok(None) => break,
+                Err(e) => {
+                    let err = format!("Failed to retrieve block info for block {n}: {e}");
+                    return Err(StarknetWsApiError::internal_server_error(err).into());
+                }
+            };
+
+            send_block_header(&sink, block_info, block_n).await?;
+            block_n = block_n.saturating_add(1);
+        }
+
+        // We need to check the block number at each iteration as the first
+        // time this is exectued we might already have received some blocks
+        // from the backend which we manually fecthed from db
+        loop {
+            tokio::select! {
+                block_info = rx.recv() => {
+                    let block_info = block_info.or_internal_server_error("Failed to retrieve block info")?;
+                    if block_info.header.block_number == block_n {
```

Comment:
can be added a check also to close the connection if a block is missing

Reply:
This is already the case as we will return an internal server error if we cannot retrieve the latest block from the backend block info broadcast channel, which will automatically close the connection.

> [!NOTE]
> The actual connection lifecycle is handled by `jsonrpsee` and not us: the connection is simply dropped whenever we exit the method.